File: internal/types/snapshot.go
package types

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// Snapshot represents a point-in-time backup snapshot
type Snapshot struct {
	ID        string                   `json:"id"`
	Timestamp time.Time                `json:"timestamp"`
	Files     map[string]*FileSnapshot `json:"files"`
	Message   string                   `json:"message,omitempty"`
}

// FileSnapshot represents a single file in a snapshot
type FileSnapshot struct {
	Path     string    `json:"path"`
	Hash     string    `json:"hash"`
	Size     int64     `json:"size"`
	Modified time.Time `json:"modified"`
}

// SnapshotDiff represents changes between two snapshots
type SnapshotDiff struct {
	From     string   `json:"from"`
	To       string   `json:"to"`
	Added    []string `json:"added"`
	Removed  []string `json:"removed"`
	Modified []string `json:"modified"`
}

// GenerateID generates a snapshot ID from a timestamp
// Format: yyyyMMdd-HHmmss-SSS (includes milliseconds to prevent collisions)
func GenerateID(t time.Time) string {
	// Format with milliseconds: 20060102-150405-000
	ms := t.UnixMilli() % 1000
	return fmt.Sprintf("%s-%03d", t.Format("20060102-150405"), ms)
}

// FromDirectory creates a snapshot from a directory
func FromDirectory(path string, exclude []string, message string) (*Snapshot, error) {
	return FromDirectoryWithTimestamp(path, exclude, message, time.Now())
}

// FromDirectoryWithTimestamp creates a snapshot from a directory with a specific timestamp
func FromDirectoryWithTimestamp(path string, exclude []string, message string, timestamp time.Time) (*Snapshot, error) {
	id := GenerateID(timestamp)
	files := make(map[string]*FileSnapshot)

	// Check if directory exists
	info, err := os.Stat(path)
	if err != nil {
		return nil, fmt.Errorf("directory does not exist: %s: %w", path, err)
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("path is not a directory: %s", path)
	}

	// Walk the directory tree
	err = filepath.Walk(path, func(filePath string, fileInfo os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip directories
		if fileInfo.IsDir() {
			return nil
		}

		// Get relative path
		relativePath, err := filepath.Rel(path, filePath)
		if err != nil {
			return fmt.Errorf("failed to get relative path: %w", err)
		}

		// Check exclusions
		if shouldExclude(relativePath, exclude) {
			return nil
		}

		// Create file snapshot
		fileSnapshot, err := fromFile(filePath, relativePath)
		if err != nil {
			return fmt.Errorf("failed to snapshot file %s: %w", relativePath, err)
		}

		files[relativePath] = fileSnapshot
		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("failed to walk directory: %w", err)
	}

	return &Snapshot{
		ID:        id,
		Timestamp: timestamp,
		Files:     files,
		Message:   message,
	}, nil
}

// fromFile creates a FileSnapshot from an actual file
func fromFile(filePath string, relativePath string) (*FileSnapshot, error) {
	// Open file
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	// Calculate SHA-256 hash
	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		return nil, fmt.Errorf("failed to hash file: %w", err)
	}
	hashString := fmt.Sprintf("%x", hash.Sum(nil))

	// Get file info
	fileInfo, err := file.Stat()
	if err != nil {
		return nil, fmt.Errorf("failed to stat file: %w", err)
	}

	return &FileSnapshot{
		Path:     relativePath,
		Hash:     hashString,
		Size:     fileInfo.Size(),
		Modified: fileInfo.ModTime(),
	}, nil
}

// Diff calculates the difference between this snapshot and another
func (s *Snapshot) Diff(other *Snapshot) *SnapshotDiff {
	diff := &SnapshotDiff{
		From:     other.ID,
		To:       s.ID,
		Added:    []string{},
		Removed:  []string{},
		Modified: []string{},
	}

	// Find added and modified files
	for path, file := range s.Files {
		if otherFile, exists := other.Files[path]; !exists {
			diff.Added = append(diff.Added, path)
		} else if file.Hash != otherFile.Hash {
			diff.Modified = append(diff.Modified, path)
		}
	}

	// Find removed files
	for path := range other.Files {
		if _, exists := s.Files[path]; !exists {
			diff.Removed = append(diff.Removed, path)
		}
	}

	return diff
}

// IsEmpty returns true if the diff has no changes
func (d *SnapshotDiff) IsEmpty() bool {
	return len(d.Added) == 0 && len(d.Removed) == 0 && len(d.Modified) == 0
}

// TotalChanges returns the total number of changes
func (d *SnapshotDiff) TotalChanges() int {
	return len(d.Added) + len(d.Removed) + len(d.Modified)
}

// String returns a string representation of the diff
func (d *SnapshotDiff) String() string {
	if d.IsEmpty() {
		return "No changes"
	}

	parts := []string{}
	if len(d.Added) > 0 {
		parts = append(parts, fmt.Sprintf("+%d added", len(d.Added)))
	}
	if len(d.Modified) > 0 {
		parts = append(parts, fmt.Sprintf("~%d modified", len(d.Modified)))
	}
	if len(d.Removed) > 0 {
		parts = append(parts, fmt.Sprintf("-%d removed", len(d.Removed)))
	}
	return strings.Join(parts, ", ")
}

// PrintDetailed prints a detailed view of the diff
func (d *SnapshotDiff) PrintDetailed() {
	if d.IsEmpty() {
		fmt.Println("No changes detected.")
		return
	}

	if len(d.Added) > 0 {
		fmt.Println("\n  Added:")
		for _, f := range d.Added {
			fmt.Printf("    + %s\n", f)
		}
	}
	if len(d.Modified) > 0 {
		fmt.Println("\n  Modified:")
		for _, f := range d.Modified {
			fmt.Printf("    ~ %s\n", f)
		}
	}
	if len(d.Removed) > 0 {
		fmt.Println("\n  Removed:")
		for _, f := range d.Removed {
			fmt.Printf("    - %s\n", f)
		}
	}
}

// shouldExclude checks if a path should be excluded based on patterns
func shouldExclude(path string, patterns []string) bool {
	for _, pattern := range patterns {
		if strings.HasSuffix(pattern, "/") {
			// Directory pattern
			if strings.HasPrefix(path, pattern) || strings.Contains(path, "/"+pattern) {
				return true
			}
		} else if strings.HasPrefix(pattern, "*.") {
			// Extension pattern
			if strings.HasSuffix(path, pattern[1:]) {
				return true
			}
		} else if strings.Contains(pattern, "**") {
			// Glob pattern - simplified matching
			regexPattern := strings.ReplaceAll(pattern, "**", ".*")
			regexPattern = strings.ReplaceAll(regexPattern, "*", "[^/]*")
			regex := regexp.MustCompile(regexPattern)
			if regex.MatchString(path) {
				return true
			}
		} else if path == pattern || strings.HasSuffix(path, "/"+pattern) {
			return true
		}
	}
	return false
}

// String returns a string representation of the snapshot
func (s *Snapshot) String() string {
	return fmt.Sprintf("Snapshot(%s, %d files)", s.ID, len(s.Files))
}

// ToJSON serializes the snapshot to JSON
func (s *Snapshot) ToJSON() ([]byte, error) {
	return json.Marshal(s)
}

// FromJSON deserializes a snapshot from JSON
func FromJSON(data []byte) (*Snapshot, error) {
	var snapshot Snapshot
	err := json.Unmarshal(data, &snapshot)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal snapshot: %w", err)
	}
	return &snapshot, nil
}

// MergeWithSources combines multiple snapshots into a single snapshot
// Each snapshot's files are prefixed with their source base name to avoid conflicts
// For example, files from ~/.openclaw become ".openclaw/file.txt"
func MergeWithSources(snapshots []*Snapshot, sourcePaths []string, message string, timestamp time.Time) (*Snapshot, error) {
	if len(snapshots) == 0 {
		return nil, fmt.Errorf("no snapshots to merge")
	}
	if len(snapshots) != len(sourcePaths) {
		return nil, fmt.Errorf("number of snapshots (%d) does not match number of source paths (%d)", len(snapshots), len(sourcePaths))
	}

	// Generate a single ID for the merged snapshot
	id := GenerateID(timestamp)

	merged := &Snapshot{
		ID:        id,
		Timestamp: timestamp,
		Files:     make(map[string]*FileSnapshot),
		Message:   message,
	}

	// Merge all files from all snapshots
	for i, snapshot := range snapshots {
		// Use the base name of the source path as the prefix
		sourceBase := filepath.Base(sourcePaths[i])

		for relPath, fileSnapshot := range snapshot.Files {
			// Prefix the path with source base name
			prefixedPath := filepath.Join(sourceBase, relPath)
			merged.Files[prefixedPath] = &FileSnapshot{
				Path:     prefixedPath,
				Hash:     fileSnapshot.Hash,
				Size:     fileSnapshot.Size,
				Modified: fileSnapshot.Modified,
			}
		}
	}

	return merged, nil
}

File: lib/src/backup/snapshot.dart
import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';
import 'package:intl/intl.dart';

/// Represents a point-in-time backup snapshot
class Snapshot {
  final String id;
  final DateTime timestamp;
  final Map<String, FileSnapshot> files;
  final String? message;

  Snapshot({
    required this.id,
    required this.timestamp,
    required this.files,
    this.message,
  });

  /// Generate a snapshot ID from timestamp
  static String generateId(DateTime dt) {
    return DateFormat('yyyyMMdd-HHmmss').format(dt);
  }

  /// Create a snapshot from a directory
  static Future<Snapshot> fromDirectory(
    String path, {
    List<String> exclude = const [],
    String? message,
  }) async {
    final timestamp = DateTime.now();
    final id = generateId(timestamp);
    final files = <String, FileSnapshot>{};

    final dir = Directory(path);
    if (!dir.existsSync()) {
      throw ArgumentError('Directory does not exist: $path');
    }

    await for (final entity in dir.list(recursive: true, followLinks: false)) {
      if (entity is File) {
        final relativePath = entity.path.substring(path.length + 1);
        
        // Check exclusions
        if (_shouldExclude(relativePath, exclude)) continue;

        final fileSnapshot = await FileSnapshot.fromFile(entity, relativePath);
        files[relativePath] = fileSnapshot;
      }
    }

    return Snapshot(
      id: id,
      timestamp: timestamp,
      files: files,
      message: message,
    );
  }

  /// Calculate diff between this snapshot and another
  SnapshotDiff diff(Snapshot other) {
    final added = <String>[];
    final removed = <String>[];
    final modified = <String>[];

    // Find added and modified files
    for (final path in files.keys) {
      if (!other.files.containsKey(path)) {
        added.add(path);
      } else if (files[path]!.hash != other.files[path]!.hash) {
        modified.add(path);
      }
    }

    // Find removed files
    for (final path in other.files.keys) {
      if (!files.containsKey(path)) {
        removed.add(path);
      }
    }

    return SnapshotDiff(
      from: other.id,
      to: id,
      added: added,
      removed: removed,
      modified: modified,
    );
  }

  /// Serialize to JSON
  Map<String, dynamic> toJson() => {
    'id': id,
    'timestamp': timestamp.toIso8601String(),
    'message': message,
    'files': files.map((k, v) => MapEntry(k, v.toJson())),
  };

  /// Deserialize from JSON
  factory Snapshot.fromJson(Map<String, dynamic> json) {
    final filesJson = json['files'] as Map<String, dynamic>;
    return Snapshot(
      id: json['id'] as String,
      timestamp: DateTime.parse(json['timestamp'] as String),
      message: json['message'] as String?,
      files: filesJson.map(
        (k, v) => MapEntry(k, FileSnapshot.fromJson(v as Map<String, dynamic>)),
      ),
    );
  }

  static bool _shouldExclude(String path, List<String> patterns) {
    for (final pattern in patterns) {
      if (pattern.endsWith('/')) {
        // Directory pattern
        if (path.startsWith(pattern) || path.contains('/$pattern')) {
          return true;
        }
      } else if (pattern.startsWith('*.')) {
        // Extension pattern
        if (path.endsWith(pattern.substring(1))) {
          return true;
        }
      } else if (pattern.contains('**')) {
        // Glob pattern - simplified matching
        final regex = RegExp(
          pattern.replaceAll('**/', '.*').replaceAll('*', '[^/]*'),
        );
        if (regex.hasMatch(path)) {
          return true;
        }
      } else if (path == pattern || path.endsWith('/$pattern')) {
        return true;
      }
    }
    return false;
  }

  @override
  String toString() => 'Snapshot($id, ${files.length} files)';
}

/// Represents a single file in a snapshot
class FileSnapshot {
  final String path;
  final String hash;
  final int size;
  final DateTime modified;

  FileSnapshot({
    required this.path,
    required this.hash,
    required this.size,
    required this.modified,
  });

  /// Create from an actual file
  static Future<FileSnapshot> fromFile(File file, String relativePath) async {
    final bytes = await file.readAsBytes();
    final hash = sha256.convert(bytes).toString();
    final stat = await file.stat();

    return FileSnapshot(
      path: relativePath,
      hash: hash,
      size: stat.size,
      modified: stat.modified,
    );
  }

  Map<String, dynamic> toJson() => {
    'path': path,
    'hash': hash,
    'size': size,
    'modified': modified.toIso8601String(),
  };

  factory FileSnapshot.fromJson(Map<String, dynamic> json) => FileSnapshot(
    path: json['path'] as String,
    hash: json['hash'] as String,
    size: json['size'] as int,
    modified: DateTime.parse(json['modified'] as String),
  );
}

/// Represents changes between two snapshots
class SnapshotDiff {
  final String from;
  final String to;
  final List<String> added;
  final List<String> removed;
  final List<String> modified;

  SnapshotDiff({
    required this.from,
    required this.to,
    required this.added,
    required this.removed,
    required this.modified,
  });

  bool get isEmpty => added.isEmpty && removed.isEmpty && modified.isEmpty;
  int get totalChanges => added.length + removed.length + modified.length;

  @override
  String toString() {
    if (isEmpty) return 'No changes';
    
    final parts = <String>[];
    if (added.isNotEmpty) parts.add('+${added.length} added');
    if (modified.isNotEmpty) parts.add('~${modified.length} modified');
    if (removed.isNotEmpty) parts.add('-${removed.length} removed');
    return parts.join(', ');
  }

  void printDetailed() {
    if (isEmpty) {
      print('No changes detected.');
      return;
    }

    if (added.isNotEmpty) {
      print('\n  Added:');
      for (final f in added) {
        print('    + $f');
      }
    }
    if (modified.isNotEmpty) {
      print('\n  Modified:');
      for (final f in modified) {
        print('    ~ $f');
      }
    }
    if (removed.isNotEmpty) {
      print('\n  Removed:');
      for (final f in removed) {
        print('    - $f');
      }
    }
  }
}

